holidays/countries/thailand.py

```diff
@@ -37,131 +200,634 @@ def add_holiday(dt, holiday_name):
                 return
 
             self[dt] = holiday_name
-            if self.observed and self._is_weekend(dt):
-                in_lieu = dt + td(days=2 if dt.weekday() == SAT else 1)
-                while in_lieu.year == year and in_lieu in self:
-                    in_lieu += td(days=+1)
 
-                add_holiday(in_lieu, f"{holiday_name} (in lieu)")
+            # !!! If Public Holiday falls on weekends, (in lieu) on Monday !!!
+            # Latest iteration was in 2001 (B.E. 2554)
+            # Data from 1992-1994 and 1998-2000 are declared discretely in
+            #   special_holidays declarations above.
+            # Sources:
+            #   (isranews.org 's wbm) http://tiny.cc/wa_isranews_inlieu_hist
```

Comment:
I didn't quite understand from this article (machine translated to English, of course) "in lieu" exactly rule: holiday is transferred to Monday only or to next working day?

Reply:
And New Year's Eve transfer from weekend too (to next year)?

Reply:
@KJhellico Thank's for the question; I did check with BoT's date regarding New Year's Eve (in lieu) holidays, and the lack of standard on the decision came out weirdly, at the very least...

That seems to be the case for [1996 (B.E. 2539)](https://www.bot.or.th/Thai/FinancialInstitutions/FIholiday/Pages/1996.aspx), [2001 (B.E. 2544)](https://www.bot.or.th/Thai/FinancialInstitutions/FIholiday/Pages/2001.aspx), [2004 (B.E. 2547)](https://www.bot.or.th/Thai/FinancialInstitutions/FIholiday/Pages/2004.aspx), [2007 (B.E. 2550)](https://www.bot.or.th/Thai/FinancialInstitutions/FIholiday/Pages/2007.aspx), and [2018 (B.E. 2561)](https://www.bot.or.th/Thai/FinancialInstitutions/FIholiday/Pages/2018.aspx) .

Not for [2006 (B.E. 2549)](https://www.bot.or.th/Thai/FinancialInstitutions/FIholiday/Pages/2006.aspx) and [2023 (B.E. 2566)](https://www.bot.or.th/Thai/FinancialInstitutions/FIholiday/Pages/2023.aspx), whereas it's denoted that Monday, Jan 1st is off for New Year's Eve and New Year's Day on the weekends.

[2012 (B.E. 2555)](https://www.bot.or.th/Thai/FinancialInstitutions/FIholiday/Pages/2012.aspx) and [2017 (B.E. 2560)](https://www.bot.or.th/Thai/FinancialInstitutions/FIholiday/Pages/2017.aspx) are weird cases where both New Year holidays are on the weekends and are counted as such for only Monday, but they added another day anyway on the following Tuesday "for economic reasons".

In any case, I think it would be for the best if we added them in the `special_holidays` section instead (some of these in lieu cases are already there AFAIK)

Reply:
Also, for clarification on the source - they did intentionally specified `next Monday` for 1961-1973, and `workdays` for other cases (Mon-Fri)

Otherwise, the actual wording for 2001 (B.E. 2544) [doesn't specify that New Year's Eve holidays can be applied in lieu in the following year](https://resolution.soc.go.th/?prep_id=99159317)
```
ให้ถือเป็นหลักการว่า กรณีวันหยุดราชการประจำปีวันใดตรงกับวันหยุดราชการประจำสัปดาห์
ให้เลื่อนวันหยุดราชการประจำปีวันนั้นไปหยุดในวันทำการถัดไป โดยให้หยุดชดเชยได้ไม่เกิน 1 วัน
```

The 2004 (B.E. 2547)'s wording does only add that [the government reserved the right to cancel or move in-lieu public holidays as they see fit](https://resolution.soc.go.th/?prep_id=196007)
```
เห็นควรให้ถือปฏิบัติตามมติคณะรัฐมนตรีเมื่อวันที่ 1 พฤษภาคม 2544 ที่ให้ถือเป็นหลักการว่า 
กรณีวันหยุดราชการประจำปีวันใดตรงกับวันหยุดราชการประจำสัปดาห์ ให้เลื่อนวันหยุดราชการ
ประจำปีวันนั้นไปหยุดในวันทำการถัดไปโดยให้หยุดชดเชยได้ไม่เกิน 1 วันต่อไป โดยให้
คณะรัฐมนตรีพิจารณาประกาศกำหนดให้งดวันหยุดชดเชยหรือเปลี่ยนวันหยุดชดเชยได้ตามความเหมาะสม
```

Although the subsequent meeting for [2006 (B.E. 2549)](https://resolution.soc.go.th/?prep_id=205799) and [2007 (B.E. 2550)](https://resolution.soc.go.th/?prep_id=210744) holidays does applied the New Year's Eve holiday from the previous year as in lieus for the next one.

Reply:
> Also, for clarification on the source - they did intentionally specified `next Monday` for 1961-1973, and `workdays` for other cases (Mon-Fri)

Then should we make relevant changes in `add_holiday` for 1961-73 (Monday only)?

Reply:
I'll see what I can do on this one.

Reply:
Another case we have to accommodate: If Songkran Festival's 3-day holiday happens to have 2 of its date fall on the weekends, then only 1 in lieu holiday is added.

This is only the case here since **the whole Songkran festival is counted as the same event**, but **New Year's Eve and New Year's Day are considered separate holidays**...

~~Gosh, I hate how complicated the Thai holiday system can be, on top of the Thai lunar calendar stuff~~

Reply:
> Another case we have to accommodate: If Songkran Festival's 3-day holiday happens to have 2 of its date fall on the weekends, then only 1 in lieu holiday is added.

How about Asarnha Bucha and Buddhist Lent Day? They are always consecutive.

Reply:
When two holidays fall on same day on weekend - it will be 1 day in lieu?

Reply:
> How about Asarnha Bucha and Buddhist Lent Day? They are always consecutive.

They seemed to be considered as the same holiday for [1997 (B.E. 2540)](https://www.myhora.com/%e0%b8%9b%e0%b8%8f%e0%b8%b4%e0%b8%97%e0%b8%b4%e0%b8%99/%e0%b8%9b%e0%b8%8f%e0%b8%b4%e0%b8%97%e0%b8%b4%e0%b8%99-%e0%b8%9e.%e0%b8%a8.2540.aspx), [2004 (B.E. 2547)](https://www.myhora.com/%e0%b8%9b%e0%b8%8f%e0%b8%b4%e0%b8%97%e0%b8%b4%e0%b8%99/%e0%b8%9b%e0%b8%8f%e0%b8%b4%e0%b8%97%e0%b8%b4%e0%b8%99-%e0%b8%9e.%e0%b8%a8.2547.aspx), [2017 (B.E. 2560)](https://www.myhora.com/%e0%b8%9b%e0%b8%8f%e0%b8%b4%e0%b8%97%e0%b8%b4%e0%b8%99/%e0%b8%9b%e0%b8%8f%e0%b8%b4%e0%b8%97%e0%b8%b4%e0%b8%99-%e0%b8%9e.%e0%b8%a8.2560.aspx) and [2021 (B.E 2564)](https://www.myhora.com/%e0%b8%9b%e0%b8%8f%e0%b8%b4%e0%b8%97%e0%b8%b4%e0%b8%99/%e0%b8%9b%e0%b8%8f%e0%b8%b4%e0%b8%97%e0%b8%b4%e0%b8%99-%e0%b8%9e.%e0%b8%a8.2564.aspx) - thus when both date fall on the weekends only an in lieu public holiday on Monday is issued by the government.

Reply:
Judging from the consecutive holiday cases we have so far - seems like the in-lieus are only added for the next Monday (with the exception being New Year's Eve-New Year's Day in-lieus); this would simplify our code a little bit.

In other news, I don't think I can get the code for multiple holidays being on the same day to display correctly when in-lieu happens i.e. `Father's Day, HM King Bhumibol Adulyadej's Birthday` and `HM Queen Sirikit The Queen Mother's Birthday, Mother's Day` combo) to display properly, same could be said for that one time `HM Queen Suthida's Birthday, Visakha Bucha` happened to consides with each other (they were given only 1 in lieu combined).

We may as well declare the in-lieus on each holiday instance instead of having it being added in the `add_holiday( )` section to help with future regional holidays which didn't have in-lieus available for them.